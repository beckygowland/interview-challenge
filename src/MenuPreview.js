import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MenuPreviewItem from './MenuPreviewItem';

class MenuPreview extends React.PureComponent {

    render() {
        const { selectedItems } = this.props;
        return (
            <div className="col-8">
                <h2>Menu preview</h2>
                <ul className="menu-preview">
                    {selectedItems.map((item) => (
                        <MenuPreviewItem
                            key={item.id + ':' + item.position}
                            {...item} />
                    ))}
                </ul>
            </div>
        )
    }
};

MenuPreview.propTypes = {
    selectedItems: PropTypes.array.isRequired,
};

const mapStateToProps = ({ selectedItems }) => {
    // we need a unique key for each menu item, and since the
    // same menu item can be added multiple times, the position/order
    // will be calculated in relation to the items with the same id
    let idCount = {};
    return {
        selectedItems: selectedItems.map((item) => {
            idCount[item.id] = idCount[item.id] !== undefined ? 
                idCount[item.id] + 1 : 0;
            return {
                ...item,
                position: idCount[item.id],
            }
        })
    }
};

export default connect(mapStateToProps, {})(MenuPreview);