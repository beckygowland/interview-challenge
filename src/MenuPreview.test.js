import React from 'react';
import { MenuPreview, mapStateToProps } from './MenuPreview';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders', () => {
  const div = shallow(<MenuPreview selectedItems={[]} />);
  expect(div).toMatchSnapshot();
  expect(div.find('ul').children().length).toBe(0);
});

it('renders list items', () => {
  const selectedItems = [{ id: 1, position: 0 }, { id: 2, position: 0 }];
  const div = shallow(<MenuPreview selectedItems={selectedItems} />);
  expect(div.find('ul').children().length ).toBe(2);
});

it('mapStateToProps', () => {
  const selectedItems = [
    { id: 1 },
    { id: 2 },
    { id: 1 },
  ]
  const props = mapStateToProps({ selectedItems })
  expect(props.selectedItems.length).toBe(3);
  expect(props.selectedItems[0].position).toBe(0);
  expect(props.selectedItems[1].position).toBe(0);
  expect(props.selectedItems[2].position).toBe(1);
});
