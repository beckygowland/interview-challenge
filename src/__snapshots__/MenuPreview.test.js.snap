// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuPreview
    selectedItems={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Menu preview
</h2>,
        <ul
          className="menu-preview"
>
          
</ul>,
      ],
      "className": "col-8",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Menu preview",
        },
        "ref": null,
        "rendered": "Menu preview",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "menu-preview",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Menu preview
</h2>,
          <ul
            className="menu-preview"
>
            
</ul>,
        ],
        "className": "col-8",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Menu preview",
          },
          "ref": null,
          "rendered": "Menu preview",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "menu-preview",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
